BHoM_UI/Components/Adapter/Pull.cs

```diff
@@ -80,19 +81,43 @@ public override object Run(object[] inputs)
         /**** Public Method               ****/
         /*************************************/
 
-        [Description("Pull objects from the external software")]
+        [Description("Adapter Action 'Pull': imports objects from an external software. \nAlso allows for querying through Requests.")]
         [Input("adapter", "Adapter to the external software")]
-        [Input("request", "Filter on the objects to pull")]
-        [Input("config", "Pull config")]
+        [Input("request", "Input a Request to pull only objects that satisfy a certain rule. \n" + 
+            "Input a Type to pull only objects of that type.")]
+        [Input("pullType", "Pull type. Connect the enum PullType for all the alternatives.")]
+        [Input("actionConfig", "Configuration for this Action. You can input an ActionConfig (it contains the configs common to all Toolkits); \n" +
+            "consider that Toolkits may have a custom ActionConfig (e.g. GSAConfig, SpeckleConfig).")]
         [Input("active", "Execute the pull")]
-        [Output("objects","Objects pulled")]
-        public static IEnumerable<object> Pull(BHoMAdapter adapter, IRequest request = null, Dictionary<string, object> config = null, bool active = false)
+        [Output("objects", "Objects pulled")]
+        public static IEnumerable<object> Pull(BHoMAdapter adapter, object request = null,
+            PullType pullType = PullType.AdapterDefault, ActionConfig actionConfig = null,
+            bool active = false)
         {
+            // ---------------------------------------------//
+            // Mandatory Adapter Action set-up              //
+            //----------------------------------------------//
+            // The following are mandatory set-ups to be ALWAYS performed 
+            // before the Adapter Action is called,
+            // whether the Action is overrided at the Toolkit level or not.
+
+            // If unset, set the actionConfig to a new ActionConfig.
+            actionConfig = actionConfig == null ? new ActionConfig() : actionConfig;
```

Comment:
All of this happens because we allow users to override the `Pull` method, in this case. If we prevent that, we can avoid spreading a lot of checks in other projects. This is explained in the Adapter pr https://github.com/BHoM/BHoM_Adapter/pull/164

Reply:
We need to be able to override the `Pull` in some of the lower level adapters

Reply:
I agree with both - a mechanism like the one suggested by @epignatelli in https://github.com/BHoM/BHoM_Adapter/pull/164 would still be advisable in my mind, so we can limit overrides to only the _strictly neededed_ cases.

Reply:
> so we can limit overrides to only the strictly neededed cases.

Please no! There is nothing more painful when you are hacking with someone else code than to realise that the stuff you want to change is not overridable.

Reply:
@adecler that's why I said "limit" and didn't say "disallow completely"!
Could you please confirm you would oppose a strategy where we have:
- overridable "Before[SomeActionName]" and "After[SomeActionName]" (where ActionName is Push/Pull/etc), that run before and after each individual adapter action, and
- overridable Actions (in which the override is permitted, but discouraged unless strictly needed)?

Reply:
I would. Anything that doesn't allow to completely override the adapter actions is a step back for me.  `Before` and `After` overridable methods still means that you have no say on code being executed in the middle. There are better ways to guide new contributors without limiting advanced developers than restricting access to code.